pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,448 @@
+		}
+
+		mappingIdx := loc.MappingId - 1
+		if int(mappingIdx) >= len(profile.Mapping) {
+			return nil, fmt.Errorf("invalid profile: location at index %d references non-existent mapping %d", i, loc.MappingId)
+		}
+
+		if !mappingsToSymbolize[loc.MappingId] {
+			continue
+		}
+
+		locsByMapping[loc.MappingId] = append(locsByMapping[loc.MappingId], locToSymbolize{
+			idx: i,
+			loc: loc,
+		})
+	}
+
+	return locsByMapping, nil
+}
+
+// symbolizeLocationsForMapping symbolizes a single mapping group
+func (s *Symbolizer) symbolizeLocationsForMapping(ctx context.Context, profile *googlev1.Profile, mappingID uint64, locs []locToSymbolize) error {
+	mapping := profile.Mapping[mappingID-1]
+
+	binaryName, err := s.extractBinaryName(profile, mapping)
+	if err != nil {
+		return err
+	}
+
+	buildID, err := s.extractBuildID(profile, mapping)
+	if err != nil {
+		return err
+	}
+
+	if buildID == "" {
+		return nil
+	}
+
+	req := s.createSymbolizationRequest(binaryName, buildID, mapping, locs)
+
+	if err := s.symbolize(ctx, &req); err != nil {
+		return err
+	}
+
+	s.updateProfileWithSymbols(profile, mapping, locs, req.locations)
+
+	return nil
+}
+
+// extractBinaryName extracts the binary name from the mapping
+func (s *Symbolizer) extractBinaryName(profile *googlev1.Profile, mapping *googlev1.Mapping) (string, error) {
+	if mapping.Filename < 0 || int(mapping.Filename) >= len(profile.StringTable) {
+		return "", fmt.Errorf("invalid mapping: filename index %d out of range (string table length: %d)",
+			mapping.Filename, len(profile.StringTable))
+	}
+
+	fullPath := profile.StringTable[mapping.Filename]
+	return filepath.Base(fullPath), nil
+}
+
+// extractBuildID extracts and sanitizes the build ID from the mapping
+func (s *Symbolizer) extractBuildID(profile *googlev1.Profile, mapping *googlev1.Mapping) (string, error) {
+	buildID := profile.StringTable[mapping.BuildId]
+	sanitizedBuildID, err := sanitizeBuildID(buildID)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "Invalid buildID", "buildID", buildID)
+		return "", err
+	}
+
+	return sanitizedBuildID, nil
+}
+
+// createSymbolizationRequest creates a symbolization request for a mapping group
+func (s *Symbolizer) createSymbolizationRequest(binaryName, buildID string, mapping *googlev1.Mapping, locs []locToSymbolize) request {
+	req := request{
+		buildID:    buildID,
+		binaryName: binaryName,
+		locations:  make([]*location, len(locs)),
+	}
+
+	for i, loc := range locs {
+		req.locations[i] = &location{
+			address: loc.loc.Address,
+			mapping: &pprof.Mapping{
+				Start:   mapping.MemoryStart,
+				Limit:   mapping.MemoryLimit,
+				Offset:  mapping.FileOffset,
+				BuildID: buildID,
+			},
+		}
+	}
+
+	return req
+}
+
+// updateProfileWithSymbols updates the profile with symbolization results
+func (s *Symbolizer) updateProfileWithSymbols(profile *googlev1.Profile, mapping *googlev1.Mapping, locs []locToSymbolize, symLocs []*location) {
```

Comment:
Can we reorganize the code a bit, to call this function once per pprof instead of once per mapping? Othervise we create all the maps multiple times